### PR TITLE
Implement NSString containsString methods

### PR DESCRIPTION
**Pull request by aballway**
Resovles #2057

### CODE REVIEW

File: tests/unittests/Foundation/NSStringTests.m [hunk 1]
```diff
@@ -725,3 +725,39 @@ void testUrlCharacterSetEncoding(NSString* decodedString, NSString* encodedStrin
...
+    EXPECT_TRUE([@"TÉST" localizedCaseInsensitiveContainsString:@"T"]);
+    EXPECT_TRUE([@"TÉST" localizedCaseInsensitiveContainsString:@"TÉST"]);
+    EXPECT_TRUE([@"TÉST" localizedCaseInsensitiveContainsString:@"tést"]);
+    EXPECT_FALSE([@"TÉST" localizedCaseInsensitiveContainsString:@"Y"]);
+    EXPECT_FALSE([@"TÉST" localizedCaseInsensitiveContainsString:@"TÉSTY"]);
+    EXPECT_FALSE([@"TÉST" localizedCaseInsensitiveContainsString:@"TEST"]);
+    EXPECT_FALSE([@"TÉST" localizedCaseInsensitiveContainsString:@"test"]);
+}
+
+TEST(NSString, LocalizedStandardContainsString) {
```
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSStringTests.m]: Add tests with other languages.

**Comment by DHowett-MSFT** [comment]
(and vice-versa)

**Comment by msft-Jeyaram** [comment]
Adding onto Dustin:
\u00E9 vs e\u0301 check if they qualify as contains.

**Comment by rajsesh-msft** [comment]
note: containsString is simply a wrapper around rangeOfString, which is probably the one we should test extensively.

File: Frameworks/Foundation/NSString.mm [hunk 2]
```diff
@@ -1120,6 +1108,44 @@ - (NSRange)rangeOfString:(NSString*)searchString options:(NSStringCompareOptions
...
+    return [self rangeOfString:str
+                       options:(NSCaseInsensitiveSearch | NSDiacriticInsensitiveSearch)
+                         range:NSMakeRange(0, [self length])
+                        locale:[NSLocale currentLocale]];
+}
+
+/**
+ @Status Interoperable
+*/
+- (BOOL)containsString:(NSString*)str {
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSString.mm]: Add test for Empty String contains NULL.

File: tests/unittests/Foundation/NSStringTests.m [hunk 3]
```diff
@@ -739,22 +744,32 @@ void testUrlCharacterSetEncoding(NSString* decodedString, NSString* encodedStrin
...
 }
 
 TEST(NSString, LocalizedStandardContainsString) {
+    EXPECT_FALSE([@"" localizedStandardContainsString:nil]);
     EXPECT_FALSE([@"" localizedStandardContainsString:@""]);
     EXPECT_FALSE([@"TÉST" localizedStandardContainsString:@""]);
     EXPECT_TRUE([@"TÉST" localizedStandardContainsString:@"T"]);
     EXPECT_TRUE([@"TÉST" localizedStandardContainsString:@"TÉST"]);
+    EXPECT_TRUE([@"TE\u0301ST" localizedStandardContainsString:@"TÉST"]);
+    EXPECT_TRUE([@"TÉST" localizedStandardContainsString:@"TE\u0301ST"]);
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSStringTests.m]: `localizedStandardContainsString:` should be case- AND diacritic-insensitive; can you test `TÉST` with its fully lowercase version, and also `te\u0301st` (which will test both diacritic sensitivity and case sensitivity?)

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSStringTests.m]: Sorry, just the decomposed different-case one; you have the first.

**Comment by aballway** [comment]
Running the tests on the reference platform, there are two differences:

1. containsString:nil with throw an exceptions
2.  On the reference platform [@"TE\u0301ST" containsString:@"TÉST"] and vice versa, whereas our platform only has them containing with diacritic-insensitive

File: tests/unittests/Foundation/NSStringTests.m [hunk 4]
```diff
@@ -727,10 +727,15 @@ void testUrlCharacterSetEncoding(NSString* decodedString, NSString* encodedStrin
...
 
 TEST(NSString, ContainsString) {
+    EXPECT_FALSE([@"" containsString:nil]);
     EXPECT_FALSE([@"" containsString:@""]);
     EXPECT_FALSE([@"TÉST" containsString:@""]);
     EXPECT_TRUE([@"TÉST" containsString:@"T"]);
     EXPECT_TRUE([@"TÉST" containsString:@"TÉST"]);
+    EXPECT_FALSE([@"TE\u0301ST" containsString:@"TÉST"]);
+    EXPECT_FALSE([@"TÉST" containsString:@"TE\u0301ST"]);
+    EXPECT_TRUE([@"اختبار النص" containsString:@"اختبار النص"]);
```
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSStringTests.m]: could you add Chinese (simplified) too?

发短信一些字符串 (texting some string)

Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSStringTests.m]: OR /*I love Microsoft*/ "我爱微软" ;)

**Comment by DHowett-MSFT** [comment]
I wonder if constant strings on the reference platform are stored in _NFC_ or _NFD_; fully normalized and canonicalized.

File: tests/unittests/Foundation/NSStringTests.m (hunk 4 above)
Comment (aballway) [inline tests/unittests/Foundation/NSStringTests.m]: I don't think we should be testing containsString's ability to translate #ByDesign

File: Frameworks/Foundation/NSString.mm [hunk 5]
```diff
@@ -1088,18 +1090,21 @@ - (NSRange)rangeOfString:(NSString*)searchString options:(NSStringCompareOptions
         return matchedRange;
     }
 
-    if (range.length == 0 || findStringLength == 0) { // ??? This last item can't be here for correct Unicode compares
-        return NSMakeRange(NSNotFound, 0);
-    }
+    if (range.length > 0 && findStringLength > 0) {
+        if ((mask & NSLiteralSearch) == 0) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSString.mm]: the early return almost certainly _improves_ code clarity here. 📏

Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSString.mm]: +1 Please update to use early returns, it makes it much easier to read and improves quality.
